File: Algorithms/1235_Maximum_Profit_in_Job_Scheduling/Python/Maximum_Profit_in_Job_Scheduling_Solution_1.py

# Space: O(n)
# Time: O(n^2)

import collections


class Solution:
    def jobScheduling(self, startTime, endTime, profit):
        cache = [0 for _ in range(max(endTime) + 1)]

        # end_time : [[start_time, profit],[start_time, profit]]
        cache_dict = collections.defaultdict(list)
        for i in range(len(endTime)):
            cache_dict[endTime[i]].append([startTime[i], profit[i]])

        for i in range(1, len(cache)):

            if i not in cache_dict:
                cache[i] = cache[i - 1]
            else:
                temp_list = cache_dict[i]
                for start_time, profit in temp_list:
                    cache[i] = max(cache[i], profit + cache[start_time])
                cache[i] = max(cache[i], cache[i - 1])
        return cache[-1]

File: Algorithms/1235_Maximum_Profit_in_Job_Scheduling/Java/MaximumProfitinJobSchedulingSolution1.java

// Space: O(n)
// Time: O(n^2)

class Solution {
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        int[] dp = new int[Arrays.stream(endTime).max().getAsInt() + 1];
        Arrays.fill(dp, 0);
        Map<Integer, List<Integer[]>> cache = new HashMap<>();
        for (int i = 0; i < endTime.length; i++) {
            cache.computeIfAbsent(endTime[i], k -> new ArrayList<>()).add(new Integer[]{startTime[i], profit[i]});
        }

        for (int i = 1; i < dp.length; i++) {
            if (cache.containsKey(i)) {
                for (Integer[] value : cache.get(i)) {
                    dp[i] = Math.max(dp[i], value[1] + dp[value[0]]);
                }
            }
            dp[i] = Math.max(dp[i], dp[i - 1]);
        }
        return dp[dp.length-1];
    }
}




